File: PopularItem.cs
//Find popular item in sorted array of natural numbers. 
//An item is popular is if its repeated n/4 times or more. 
//For Ex: 
//Input: 123444445678 
//Popular item is 4. 
//Liner scan is O(n), but solution needs to be in O(logN) complexity and O(1) space complexity.


public sealed class PopularItem
{
	public static void Main(String[] args)
	{
		
	}
	
	public static int Get(int[] items, int popScore)
	{
		if(items == null || items.Count <=0)
		{
			throw new ArgumentNullException("Argument 'items' is null");
		}
		
		if(popScore > items.Length)
		{
			return -1;
		}
		
		return Get(items, popScore, 0, items.Length-1);
	}
	
	public static int Get(int[] items, popScore, int left, int right)
	{	
		if(right < left)
		{	
			return -1;	
		}
		
		int mid = (left + right) / 2;
		int current = items[mid];
		
		int iterator = mid - popScore + 1;
		
		while(iterator <= mid)
		{
			if(items[mid] == items[iterator] && items[mid] ==items[iterator + popScore - 1])
			{
				return items[mid];
			}
			iterator++;
		}
		
		
	}
	
}

File: PopularItem.java
public class PopularItem
{
    public static void main(String[] args)
    {
	int[] arr = new int[] {1,2,2,3,4,4,4,5,6,7,8,9};
	int pop = getPopularItem(arr);
	System.out.println(pop);
    }
    
    //Given a sorted array of natural numbers, find the number which is popular.
    //Popular item is one thats present n/4 or more times in the array. n is # of items.
    //1,2,2,3,4,5,6,6,6,7
    public static int getPopularItem(int[] items){
        if(items == null){
	    return -1;
	}

	int n = items.length;
	
        if(n == 0){
	    return -1;
	}

	int popularity = n / 4;
	
	int l = items[0];
	int r = items[popularity-1];

	if(items[l] == items[r]){
	    return items[l];
	}

	r = items[n-1];
	l = items[n-popularity];
	if(items[l] == items[r]){
	    return items[l];
	}

	l = 0;
	int i = popularity;
	r = i-1;
	
	while(i <= 3*popularity){
	    int item = items[i];
	    int found = i;
	    while(l<=r){
		int mid  = (l+r)/2;
		if(items[mid] == item){
		    found = mid;
		    r = mid-1;
		}
		else{
		    l = mid+1;
		}
	    }

	    int next = found + popularity -1;
	    if(items[next] == item){
		return item;
	    }	    
	    l = i+1;
	    i = i+i;
	    r = i-1;
	   
	}
	return -1;
    }
}
